### ISSUE TITLE
Packer aborts if symlinks found locally as resource inputs

### ISSUE THREAD

**Issue by tb3088** [hashicorp/packer#4545]
Just as an example, the puppet-masterless Provisioner (like all other provisioners and such) have a verification section which tries to check if the specified manifest_file or module_path(s) element is a file or directory as applicable. However none of these top-level resource checks handle symlinks which is quite annoying and unnecessary when the SSH communicator's scpUploadDir() deals intelligently with the matter.

No, this is not to be confused with the multitude of previously raised issues about symlink preservation (or not) behavior of uploadDirectory() or uploadFile() as they process child objects. (aside: IMO all symlinks should be unrolled when copying files from one system to another). 

This is about being tolerant of usual and customary behavior by Packer Users as they organize their provisioning environments. If my 'var-files' and template JSON files can be symlinks, then so can references to ISO(s), manifest_file(s), module_path(s) and other assets that are read or evaluated locally. For another example of unnecessary breakage see  #3437.

Basically ~all calls to Stat(<item>) should be replaced with `Stat(filepath.EvalSymlinks(<item>))` and then the tests like isDir() and can be applied.

**Comment by tb3088** [comment]
A related issue is the aforementioned uploadFile()'s failure to handle Files that are Symlinks and to unroll them instead of resorting to copying the file into a temporary location and uploading that.

**Comment by tanmay-hc** [comment]
@tb3088 I believe this issue is fixed now. I tried using a symlink file and Packer is correctly able to execute the same.

```
packer {
  required_plugins {
    docker = {
      version = ">= 1.0.8"
      source  = "github.com/hashicorp/docker"
    }
  }
}

source "docker" "example" {
  image  = "ubuntu:jammy"
  commit = true
  discard = false
}

build {
  sources = ["source.docker.example"]

  provisioner "shell" {
    script = "sym.sh"
  }
}
```

Here, the file `sym.sh` is a symlink with content `echo "Hello World!"`.
Following are the logs, which are correctly printing the statement.
```
docker.example: output will be in this color.

==> docker.example: Creating a temporary directory for sharing data...
==> docker.example: Pulling Docker image: ubuntu:jammy
    docker.example: jammy: Pulling from library/ubuntu
    docker.example: Digest: sha256:ed1544e454989078f5dec1bfdabd8c5cc9c48e0705d07b678ab6ae3fb61952d2
    docker.example: Status: Image is up to date for ubuntu:jammy
    docker.example: docker.io/library/ubuntu:jammy
==> docker.example: Starting docker container...
    docker.example: Run command: docker run -v /Users/tanmay/.packer.d/tmp957536074:/packer-files -d -i -t --entrypoint=/bin/sh -- ubuntu:jammy
    docker.example: Container ID: 0d90087673c8e9a4df0ec531255ad2edb2f90034141f1a0cdf2152ddf74dfbee
==> docker.example: Using docker communicator to connect: 172.17.0.2
==> docker.example: Provisioning with shell script: sym.sh
    docker.example: Hello World!
==> docker.example: Committing the container
    docker.example: Image ID: sha256:acccf14e3f3c56e96932a44911828ef3928ae67b020ec168d77d136c103a70ea
==> docker.example: Killing the container: 0d90087673c8e9a4df0ec531255ad2edb2f90034141f1a0cdf2152ddf74dfbee
Build 'docker.example' finished after 4 seconds 726 milliseconds.

==> Wait completed after 4 seconds 726 milliseconds

==> Builds finished. The artifacts of successful builds are:
--> docker.example: Imported Docker image: sha256:acccf14e3f3c56e96932a44911828ef3928ae67b020ec168d77d136c103a70ea
```

Please let us know if you're still facing the issue.

**Comment by tanmay-hc** [comment]
Closing this issue for now. Please create a new issue if you face a problem.
